File: src/particle_system.py
"""
Particle System - Visual effects for memory healing
"""

import pygame
import random
import math
from typing import List, Tuple


class Particle:
    """Individual particle for visual effects"""
    
    def __init__(self, x: float, y: float, color: Tuple[int, int, int]):
        self.x = x
        self.y = y
        self.color = color
        
        # Random velocity
        angle = random.random() * math.pi * 2
        speed = random.uniform(20, 80)
        self.vx = math.cos(angle) * speed
        self.vy = math.sin(angle) * speed
        
        # Properties
        self.lifetime = random.uniform(0.5, 1.5)
        self.max_lifetime = self.lifetime
        self.size = random.uniform(2, 5)
        
    def update(self, dt: float) -> bool:
        """Update particle position and lifetime. Returns False when dead."""
        self.x += self.vx * dt
        self.y += self.vy * dt
        
        # Apply some gravity/float effect
        self.vy += -20 * dt  # Slight upward float
        
        # Fade out
        self.lifetime -= dt
        
        # Slow down
        self.vx *= 0.98
        self.vy *= 0.98
        
        return self.lifetime > 0
        
    def draw(self, screen: pygame.Surface):
        """Draw the particle"""
        # Calculate alpha based on lifetime
        alpha = self.lifetime / self.max_lifetime
        
        # Draw with fade
        if alpha > 0:
            # Calculate current size
            current_size = self.size * alpha
            
            # Draw glow effect
            for i in range(3):
                glow_size = current_size * (3 - i)
                glow_alpha = alpha * (0.3 / (i + 1))
                glow_color = (
                    min(255, self.color[0] + 50),
                    min(255, self.color[1] + 50),
                    min(255, self.color[2] + 50)
                )
                
                if glow_size > 0:
                    pygame.draw.circle(
                        screen,
                        glow_color,
                        (int(self.x), int(self.y)),
                        int(glow_size),
                        0
                    )
            
            # Draw core
            pygame.draw.circle(
                screen,
                self.color,
                (int(self.x), int(self.y)),
                int(current_size),
                0
            )


class ParticleSystem:
    """Manages all particles in the game"""
    
    def __init__(self, screen_width: int, screen_height: int):
        self.particles: List[Particle] = []
        self.screen_width = screen_width
        self.screen_height = screen_height
        
    def emit(self, x: float, y: float, count: int = 5, 
             color: Tuple[int, int, int] = (255, 255, 255)):
        """Emit new particles at a position"""
        for _ in range(count):
            # Add some position variance
            px = x + random.uniform(-10, 10)
            py = y + random.uniform(-10, 10)
            
            # Vary the color slightly
            color_variance = 20
            varied_color = (
                max(0, min(255, color[0] + random.randint(-color_variance, color_variance))),
                max(0, min(255, color[1] + random.randint(-color_variance, color_variance))),
                max(0, min(255, color[2] + random.randint(-color_variance, color_variance)))
            )
            
            self.particles.append(Particle(px, py, varied_color))
            
    def update(self, dt: float):
        """Update all particles"""
        # Update particles and remove dead ones
        self.particles = [p for p in self.particles if p.update(dt)]
        
        # Limit total particles to prevent performance issues
        if len(self.particles) > 500:
            self.particles = self.particles[-500:]
            
    def draw(self, screen: pygame.Surface):
        """Draw all particles"""
        # Create a temporary surface for additive blending
        temp_surface = pygame.Surface((self.screen_width, self.screen_height))
        temp_surface.fill((0, 0, 0))
        
        # Draw particles to temp surface
        for particle in self.particles:
            particle.draw(temp_surface)
            
        # Blit with additive blending for glow effect
        screen.blit(temp_surface, (0, 0), special_flags=pygame.BLEND_ADD)
        
    def clear(self):
        """Remove all particles"""
        self.particles.clear()
        
    def active_count(self) -> int:
        """Get number of active particles"""
        return len(self.particles)

File: renderer/js/particle-system.js
// Particle System for Memory Drift

class Particle {
    constructor(x, y, color) {
        this.position = new Vector2(x, y);
        this.velocity = Vector2.fromAngle(
            Utils.random(0, Math.PI * 2),
            Utils.random(20, 80)
        );
        this.color = color;
        this.lifetime = Utils.random(0.5, 1.5);
        this.maxLifetime = this.lifetime;
        this.size = Utils.random(2, 5);
        this.alpha = 1;
    }
    
    update(deltaTime) {
        // Update position
        this.position = this.position.add(this.velocity.multiply(deltaTime));
        
        // Apply slight upward drift
        this.velocity.y -= 20 * deltaTime;
        
        // Apply drag
        this.velocity = this.velocity.multiply(0.98);
        
        // Update lifetime
        this.lifetime -= deltaTime;
        this.alpha = this.lifetime / this.maxLifetime;
        
        return this.lifetime > 0;
    }
    
    draw(ctx) {
        if (this.alpha <= 0) return;
        
        const currentSize = this.size * this.alpha;
        
        // Save context
        ctx.save();
        
        // Set global alpha
        ctx.globalAlpha = this.alpha;
        
        // Draw glow effect
        for (let i = 0; i < 3; i++) {
            const glowSize = currentSize * (3 - i);
            const glowAlpha = this.alpha * (0.3 / (i + 1));
            
            ctx.globalAlpha = glowAlpha;
            ctx.fillStyle = `rgb(${Math.min(255, this.color.r + 50)}, ${Math.min(255, this.color.g + 50)}, ${Math.min(255, this.color.b + 50)})`;
            
            ctx.beginPath();
            ctx.arc(this.position.x, this.position.y, glowSize, 0, Math.PI * 2);
            ctx.fill();
        }
        
        // Draw core
        ctx.globalAlpha = this.alpha;
        ctx.fillStyle = Utils.colorToString(this.color);
        ctx.beginPath();
        ctx.arc(this.position.x, this.position.y, currentSize, 0, Math.PI * 2);
        ctx.fill();
        
        // Restore context
        ctx.restore();
    }
}

class ParticleSystem {
    constructor() {
        this.particles = [];
        this.maxParticles = 500;
    }
    
    emit(x, y, count = 5, color = { r: 255, g: 255, b: 255 }) {
        for (let i = 0; i < count; i++) {
            // Add position variance
            const px = x + Utils.random(-10, 10);
            const py = y + Utils.random(-10, 10);
            
            // Add color variance
            const colorVariance = 20;
            const variedColor = {
                r: Utils.clamp(color.r + Utils.random(-colorVariance, colorVariance), 0, 255),
                g: Utils.clamp(color.g + Utils.random(-colorVariance, colorVariance), 0, 255),
                b: Utils.clamp(color.b + Utils.random(-colorVariance, colorVariance), 0, 255)
            };
            
            this.particles.push(new Particle(px, py, variedColor));
        }
        
        // Limit particle count
        if (this.particles.length > this.maxParticles) {
            this.particles = this.particles.slice(-this.maxParticles);
        }
    }
    
    update(deltaTime) {
        // Update particles and remove dead ones
        this.particles = this.particles.filter(particle => particle.update(deltaTime));
    }
    
    draw(ctx) {
        // Use additive blending for glow effect
        const oldCompositeOperation = ctx.globalCompositeOperation;
        ctx.globalCompositeOperation = 'lighter';
        
        this.particles.forEach(particle => {
            particle.draw(ctx);
        });
        
        // Restore composite operation
        ctx.globalCompositeOperation = oldCompositeOperation;
    }
    
    clear() {
        this.particles = [];
    }
    
    getActiveCount() {
        return this.particles.length;
    }
}